### PR TITLE
Refactor Technique class

### PR DESCRIPTION
**Pull request by JaskRendix**
Following #1364 related to the refactor of the technique class (#937)

Issues:
~~(1) Now when a battle starts it shows me "pound" as techniques (self.name = "Pound") (screenshot)
Maybe we need to add a "for" in combat, but at the moment I don't know if it's a silly thing or not.
Or maybe getting rid of self.session in technique.py and fixing the parser, but at the moment I have no clue.~~

~~((2) If switching monster: crash -> see question "meta"~~(

A series of questions related to the refactoring:
1. self.category for techniques, it looks like has been created only for supporting the old system, I guess it can be deleted once the next system is applied. Is it correct?
2. self.conditions for techniques, it was planned in the beginning? If yes, then I have a PR ready to integrate the empty field inside the technique JSONs as well as in db.py.
3. what is "meta"? Do we need to create the meta.py? (as effect in status_faint and run)

### CODE REVIEW

File: tuxemon/technique/effects/status.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        already_applied = any(t for t in target.status if t.slug == slug)
+        success = (
+            not already_applied
+            and base.can_apply_status
+            and base.potency >= random.random(),
+        )
+        tech = None
+        if success:
+            if slug == "status_faint":
+                target.apply_status(base("status_faint"))
```
Comment (kerizane) [inline tuxemon/technique/effects/status.py]: Suspicious

Comment (JaskRendix) [reply tuxemon/technique/effects/status.py]: Agree, I'll update as lifeleech
```
tech = base(slug, carrier=target, link=self.user)
target.apply_status(tech)
```

Comment (JaskRendix) [reply tuxemon/technique/effects/status.py]: or maybe only target, but in that case we can remove the if about status_faint, what do you think?

Comment (kerizane) [reply tuxemon/technique/effects/status.py]: Surely it needs to be `Technique`, not `base`?

File: tuxemon/technique/effects/damage.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    Returns:
+        Dict summarizing the result.
+
+    """
+
+    name = "damage"
+    param_class = DamageEffectParameters
+
+    def apply(self, target: Monster) -> DamageEffectResult:
```
Comment (kerizane) [inline tuxemon/technique/effects/damage.py]: You need to pass the slug in somehow and load the actual technique.

Comment (kerizane) [reply tuxemon/technique/effects/damage.py]: Same for all effects.

**Comment by kerizane** [comment]
> (1) Now when a battle starts it shows me "pound" as techniques (self.name = "Pound") (screenshot)

This is because you're not loading the data for the technique that was used, you're loading the default data, which is Pound.

> (2) If switching monster: crash -> see question "meta"

meta is for techniques that aren't strictly techniques. I don't think we need meta.py, just need the code to not crash.




> * self.category for techniques, it looks like has been created only for supporting the old system, I guess it can be deleted once the next system is applied. Is it correct?

yeah, delete it


> * self.conditions for techniques, it was planned in the beginning? If yes, then I have a PR ready to integrate the empty field inside the technique JSONs as well as in db.py.

I think these will be useful later, but will need some refactoring. I will decide what to do with them later.

**Comment by JaskRendix** [comment]
Moves are visible, no crash in swapping, status_faint works (I got teleported by losing against a wild monster).
I didn't touch category at the moment, I'll do in another PR.
I removed the two "meta" from run and status_faint.

**Comment by kerizane** [comment]
Please test every status.

**Comment by JaskRendix** [comment]
So, this is the situation:
1. **poison** and **faint** and **recover** work (but if the statuses are active -> set monster status), there is no way of getting poisoned (points 4/5).
2. **status_overfeed** doesn't exist as JSON, but there is **simple_overfeed** in formula.py, so I guess someone wanted to create it, but the job isn't complete (like hardshell). Do I need to create it?
3. I assigned the **status_lifeleech** to a monster and I went in a battle, it crashed:
```
Tuxemon/tuxemon/formula.py", line 195, in simple_lifeleech
    damage = min(target.hp // 2, target.current_hp, user.hp - user.current_hp)
AttributeError: 'NoneType' object has no attribute 'hp'
```
I think it's the user in lifeleech.py, but I haven't looked into at the moment.

4. There is no way of getting **status_recover**, **status_overfeed** or **status_lifeleech**. There is no technique with these effects as well as **recover**, **lifeleech** and **overfeed**.
5. As previously said (4), look at this:

Do I need to add **status status_poison**? Because right now (before too) there is/was no way of getting poisoned during the battle.

6. we need to move a bit to the right the symbol and reduce the space between these (screenshots).
```
    set_monster_status 0,status_poison
    set_monster_status 0,status_recover
    set_monster_status 0,status_lifeleech
```

7. "Faint" in status_faint? Inside the JSON status_faint there is "Faint" where normally there is wood, metal, etc. I guess it's a mistake. I didn't find trace about this type.

Other **status_XXX** don't have the field types, while **run** has **types": []**. Do I need to standardize everything with **types": []**?

9. There is a bool called **can_apply_status**, it appears only in 4 places (overfeed json - set on true -, technique.py - set on false -, damage.py and status.py - last two as success condition -); I would like to remove the field, I guess it was inserted only for overfeed, but by creating the status_overfeed JSON we can standardize and remove a condition that can impact other statuses (since it's in damage.py and status.py).

**Comment by kerizane** [comment]
I am pretty sure poison used to work, I remember getting poisoned by cataspike.

**Comment by kerizane** [comment]
> status_overfeed doesn't exist as JSON, but there is simple_overfeed in formula.py, so I guess someone wanted to create it, but the job isn't complete (like hardshell). Do I need to create it?

Confirm the behaviour of "overfeed" (technique owned by aardart) hasn't changed. If it hasn't then create a ticket to investigate overfeed.

> There is no way of getting status_recover, status_overfeed or status_lifeleech. There is no technique with these effects as well as recover, lifeleech and overfeed.

Create a ticket.

> we need to move a bit to the right the symbol and reduce the space between these (screenshots).
    set_monster_status 0,status_poison
    set_monster_status 0,status_recover
    set_monster_status 0,status_lifeleech

Create a ticket if this is new.

> "Faint" in status_faint? 

Remove.

> Do I need to standardize everything with types": []?

Just remove types from run.

>There is a bool called can_apply_status

This logic is used as part of setting poison. Remove from overfeed.json leave it in everywhere else.

File: mods/tuxemon/db/technique/status_poison.json [hunk 3]
```diff
@@ -10,7 +10,8 @@
   "category": "poison",
   "sort": "meta",
   "effects": [
-    "status"
+    "poison",
+    "status status_poison"
```
Comment (kerizane) [inline mods/tuxemon/db/technique/status_poison.json]: `"status status_poison"`
This shouldn't be here, it should be on the techniques "suck_poison" and "venemous_tentacle".

**Review by kerizane — CHANGES_REQUESTED** [review]

File: mods/tuxemon/db/technique/status_poison.json (hunk 3 above)
Comment (kerizane) [reply mods/tuxemon/db/technique/status_poison.json]: Check all the techniques that should apply a status.

Comment (JaskRendix) [reply mods/tuxemon/db/technique/status_poison.json]: Yeah, you know, that's what I wanted to say. Thanks for repeating it.

Comment (kerizane) [reply mods/tuxemon/db/technique/status_poison.json]: Also, you should remove "poison" from those techniques.

**Comment by JaskRendix** [comment]
Removed faint.
Removed types from run
Lifeleech still crashes, I need to look into.

The space between symbols isn't new, it happens on development too.
from
Screenshot from 2022-10-02 09-42-33
to
Screenshot from 2022-10-02 09-43-23

Removed can_apply_status from overfeed (screenshot where it remains)

**Comment by kerizane** [comment]
Is lifeleech the only thing left?

**Comment by JaskRendix** [comment]
Moved some parts in the #1372 since it happens the same thing in development.

**Comment by kerizane** [comment]
Is lifeleech fixed now?

Remove the conditions stuff from this PR, but leave it in another branch in your repo in case we want it later.

**Comment by JaskRendix** [comment]
No, it's not fixed lifeleech, if you have a clue about the user. It could be something silly, but I'm unable to see right now.

**Comment by JaskRendix** [comment]
Purged conditions done. Now I have a branch with conditions (included the one for the technique JSONs).

**Comment by kerizane** [comment]
Fixed

**Comment by kerizane** [comment]
I think that's it. Want me to merge?

**Comment by JaskRendix** [comment]
Sure, let's merge, then tell me if you want the draft about conditions

**Review by kerizane — APPROVED** [review]

**Pull request merged**